packages/chat-stateful-client/src/ResourceDownloadQueue.test.ts

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { CommunicationTokenCredential } from '@azure/communication-common';
+import { ChatContext } from './ChatContext';
+import { ResourceDownloadError, ResourceDownloadQueue } from './ResourceDownloadQueue';
+import { messageTemplate } from './mocks/createMockChatThreadClient';
+import { ChatMessageWithStatus } from './types/ChatMessageWithStatus';
+
+jest.mock('@azure/communication-chat');
+
+/**
+ * @private
+ */
+export const stubCommunicationTokenCredential = (): CommunicationTokenCredential => {
+  return {
+    getToken: (): Promise<{ token: string; expiresOnTimestamp: number }> => {
+      throw new Error('Not implemented');
+    },
+    dispose: (): void => {
+      /* Nothing to dispose */
+    }
+  };
+};
+
+describe('ResourceDownloadQueue api functions', () => {
+  test('should add a message to the queue and contains message', () => {
+    const context = new ChatContext();
+    const tokenCredential = stubCommunicationTokenCredential();
+    const mockMessage = messageTemplate;
+    const queue = new ResourceDownloadQueue(context, tokenCredential);
+    queue.addMessage(mockMessage);
+    expect(queue.containsMessage(mockMessage)).toBe(true);
+  });
+
+  test('start queue should call operation', async () => {
+    const context = new ChatContext();
+    const tokenCredential = stubCommunicationTokenCredential();
+    const mockMessage = messageTemplate;
+    const queue = new ResourceDownloadQueue(context, tokenCredential);
+    const operation = jest.fn();
+    queue.addMessage(mockMessage);
+    await queue.startQueue('threadId', operation);
+    expect(operation).toHaveBeenCalled();
+  });
+
+  test('adding multiple chats to queue should call operation multiple times', async () => {
+    const context = new ChatContext();
+    const tokenCredential = stubCommunicationTokenCredential();
+    const first = messageTemplate;
+    first.id = 'first';
+    const second = messageTemplate;
+    second.id = 'second';
+    const third = messageTemplate;
+    third.id = 'third';
+
+    const queue = new ResourceDownloadQueue(context, tokenCredential);
+    const operation = jest.fn();
+    queue.addMessage(first);
+    queue.addMessage(second);
+    queue.addMessage(third);
+    await queue.startQueue('threadId', operation);
+    expect(operation).toHaveBeenCalledTimes(3);
+  });
+
+  test('adding multiple chat to queue should call operation in order', async () => {
```

Comment:
Can we expand this test to make sure that only one call gets made at once?

We need to verify that the queue runs serially.

Reply:
Updated to include a new test that will validate even if the start has been called early it will still complete all operations in sequential order